torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -0,0 +1,233 @@
+        "Batch size (was ", options.batch_size_, ") ",
+        "must not be larger than the dataset size (was ",
+        dataset.size(), ")");
+    // clang-format on
+
+    if (options_.enforce_ordering_) {
+      sequencer_ =
+          torch::make_unique<OrderedSequencer<Result>>(*options_.max_jobs_);
+    } else {
+      sequencer_ = torch::make_unique<NoSequencer<Result>>();
+    }
+
+    if (options_.workers_ == 0) {
+      dataset_ = std::move(dataset);
+    }
+
+    for (size_t w = 0; w < options_.workers_; ++w) {
+      // Here we copy the dataset into the worker thread closure. Each worker
+      // has its own copy of the dataset. This means the dataset must be
+      // trivially copiable, or else we don't expect more than one worker to
+      // be in use.
+      workers_.emplace_back(
+          [this, dataset] { this->worker_thread(std::move(dataset)); });
+    }
+
+    if (!options_.defer_prefetch_) {
+      prefetch();
+    }
+  }
+
+  ~DataLoader() {
+    join();
+  }
+
+  void reset(bool prefetch = true) {
+    shuttle_.drain();
+    sampler_.reset();
+    sequence_number_ = 0;
+    if (prefetch) {
+      this->prefetch();
+    }
+  }
+
+  void prefetch(size_t requested_jobs) {
+    while (requested_jobs-- > 0) {
+      if (auto index_batch = get_index_batch()) {
+        push_job(std::move(*index_batch));
+      } else {
+        shuttle_.exhausted();
+      }
+    }
+  }
+
+  void prefetch() {
+    prefetch(*options_.max_jobs_);
+  }
+
+  at::optional<Batch> next() {
+    at::optional<Batch> batch;
+    if (options_.workers_ > 0) {
+      // sequencer_->next(...).map(|r| r.batch)
+      auto result = sequencer_->next(
+          [this] { return shuttle_.pop_result(options_.timeout_); });
+      if (result) {
+        batch = std::move(result->batch);
+      }
+      prefetch(1);
+    } else if (auto index_batch = get_index_batch()) {
+      batch = dataset_->batch(std::move(*index_batch));
+    }
+    return batch;
+  }
+
+  template <typename LoopFunction>
+  void loop(LoopFunction function) {
+    while (auto maybe_batch = next()) {
+      function(*maybe_batch);
+    }
+    reset();
+  }
+
+  void join() {
+    if (joined_) {
+      return;
+    }
+    shuttle_.drain();
+    // Send one 'quit' message per worker. Since a worker dies (exits its
+    // thread) after receiving this message, each `QuitWorker()` message will be
+    // read by exactly one worker.
+    for (size_t w = 0; w < options_.workers_; ++w) {
+      push_job(QuitWorker());
+    }
+    for (auto& worker : workers_) {
+      worker.join();
+    }
+    joined_ = true;
+  }
+
+  const DataLoaderOptions& options() const noexcept {
+    return options_;
+  }
+
+  size_t dataset_size() const noexcept {
+    return dataset_size_;
+  }
+
+ private:
+  struct Sequenced {
+    Sequenced() = default;
+    Sequenced(size_t sqn) : sequence_number(sqn) {}
+    size_t sequence_number;
+  };
+
+  struct QuitWorker {};
+  // Job = Enum(QuitWorker, IndexBatch)
```

Comment:
it's really a union/sum type, not an enum

Reply:
`Enum` is the name for sum types in OCaml and Rust

Reply:
Ugh who came up with that!